ambry-store/src/main/java/com.github.ambry.store/BlobStore.java

```diff
@@ -483,6 +486,87 @@ public void delete(MessageWriteSet messageSetToDelete) throws StoreException {
+                    + value.getAccountId() + " ContainerId in store: " + value.getContainerId(),
+                StoreErrorCodes.Authorization_Failure);
+          } else {
+            logger.warn("UPDATE authorization failure. Key: {} AccountId in store: {} ContainerId in store: {}",
+                info.getStoreKey(), value.getAccountId(), value.getContainerId());
+            metrics.ttlUpdateAuthorizationFailureCount.inc();
+          }
+        } else if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+          throw new StoreException(
+              "Cannot update TTL of " + info.getStoreKey() + " since it is already deleted in the index.",
+              StoreErrorCodes.ID_Deleted);
+        } else if (value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+          throw new StoreException("TTL of " + info.getStoreKey() + " is already updated in the index.",
+              StoreErrorCodes.Already_Updated);
+        } else if (value.getExpiresAtMs() != Utils.Infinite_Time
+            && value.getExpiresAtMs() < info.getOperationTimeMs() + ttlUpdateBufferTimeMs) {
+          throw new StoreException(
+              "TTL of " + info.getStoreKey() + " cannot be updated because it is too close to expiry. Op time (ms): "
+                  + info.getOperationTimeMs() + ". ExpiresAtMs: " + value.getExpiresAtMs(),
+              StoreErrorCodes.Update_Not_Allowed);
+        }
+      }
+      synchronized (storeWriteLock) {
+        Offset currentIndexEndOffset = index.getCurrentEndOffset();
+        if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
+          FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
+          for (MessageInfo info : infoList) {
+            IndexValue value =
+                index.findKey(info.getStoreKey(), fileSpan, EnumSet.allOf(PersistentIndex.IndexEntryType.class));
+            if (value != null) {
+              if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+                throw new StoreException(
+                    "Cannot update TTL of " + info.getStoreKey() + " since it is already deleted in the index.",
+                    StoreErrorCodes.ID_Deleted);
+              } else if (value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+                throw new StoreException("TTL of " + info.getStoreKey() + " is already updated in the index.",
+                    StoreErrorCodes.Already_Updated);
+              }
+            }
+          }
+        }
+        Offset endOffsetOfLastMessage = log.getEndOffset();
+        messageSetToUpdate.writeTo(log);
+        logger.trace("Store : {} ttl update mark written to log", dataDir);
+        for (MessageInfo info : infoList) {
+          FileSpan fileSpan = log.getFileSpanForMessage(endOffsetOfLastMessage, info.getSize());
+          index.markAsPermanent(info.getStoreKey(), fileSpan, info.getOperationTimeMs());
```

Comment:
This is a comment to `markAsPermanent()`. 

It's a little counterintuitive:  `markAsPermanent()` is calling findKey() only with `EnumSet.of(PUT, DELETE)`, but it's possible to return a key whose flag has `Ttl_Update_Index` and then throw exception. 

Also, curious about how  `IndexType.TTL_UPDATE` would be used. Didn't see it was used in this PR and #978.

Reply:
Need to understand your question better - will clarify offline